src/angular-modal-service.js

```diff
@@ -115,25 +109,35 @@ module.factory('ModalService', ['$animate', '$document', '$compile', '$controlle
             scope: modalScope,
             element: modalElement,
             close: closeDeferred.promise,
-            closed: closedDeferred.promise
+            closed: closedDeferred.promise,
+            closeModal: closeModal
           };
 
           //  ...which is passed to the caller via the promise.
           deferred.resolve(modal);
 
+          function closeModal(result, delay) {
+            if (delay === undefined || delay === null) delay = 0;
+            $timeout(function() {
+
+              cleanUpClose(result);
+
+            }, delay);
+          }
+
           function cleanUpClose(result) {
 
             //  Resolve the 'close' promise.
-            closeDeferred.resolve(result);
+            closeDeferred && closeDeferred.resolve(result);
 
             //  Let angular remove the element and wait for animations to finish.
             $animate.leave(modalElement)
                     .then(function () {
                       //  Resolve the 'closed' promise.
-                      closedDeferred.resolve(result);
+                      closedDeferred && closedDeferred.resolve(result);
```

Comment:
Didn't we want to consider rejecting the modal's promise if `cleanUpClose` was triggered by `locationChangeSuccess`?